feat: add doc for typescript auth helpers (#1368)
vaadin/flow-and-components-documentation: documentation/typescript/adding-custom-login-form-with-spring-security.asciidoc

  constructor(){
    super();
    this.onSuccess = () => {
      Router.go(this.returnUrl);
    };
  }

  render() {
    return html`
      <vaadin-login-overlay
        ?opened="${this.open}" 
        .error=${this.error}
        @login="${this.login}">    
      </vaadin-login-overlay>
    `;
  }

  async login(event: CustomEvent): Promise<LoginResult> {
    // use the Vaadin provided login helper method to obtain the LoginResult
    const result = await login(event.detail.username, event.detail.password);
    this.error = result.error;
  
    if (!result.error) {
      this.onSuccess(result);
    }

    return result;
  }

  onAfterEnter(location: RouterLocation) {
    this.returnUrl = location.redirectFrom || this.returnUrl;
  }
}
----
After the login view is defined, you should also define a route for the login view component, e.g., in the `index.ts` file. Note, the `path` for the login view component should be the same as what you defined in `SecurityConfig` for `http.formLogin().loginPage()`.
.frontend/index.ts
[source, typescript]
----
const routes = [
  {
    path: '/login',
    component: 'login-view'
  },
  // more routes
}
----

== Logout

> There two ways to do logout. You could either

Reviewer: Badly written.
```suggestion
There are two ways to implement logout. You could either:
```
Also this should probably be followed by bullets (1. and 2.) or bullet points due to how the above text ends.

Let Spring Security handle Logging out. How it works is that it handles HTTP GET requests to `/logout` and redirects to `/login?logout` in response.
For that a "Logout" button should be an regular <a> tag, i.e.
`<a href="/logout" router-ignore>Log out</a>`. The downside is that it requires a full page load.

> In order to avoid a full page reload), it would require a `/logout` route like the one below. In this case a "Logout" button should an in-app link like `<a href="/logout">Log out</a>`

Reviewer: ```suggestion
In order to avoid a full page reload the application needs to have a `/logout` route like the one below. In this case it can be triggered with a link like `<a href="/logout">Log out</a>`.
```

---
title: Adding a Custom Login Form with Spring Security
order: 91
layout: page
---

ifdef::env-github[:outfilesuffix: .asciidoc]

= Adding a Custom Login Form with Spring Security

Instead of using the <<adding-login-form-with-spring-security#, default Spring login form>>, you may want to use your own customized form. This article describes how to do that.

== Dependencies and Server Configuration

> The dependencies and server configuratiton is the same as described in <<adding-login-form-with-spring-security#, Adding a Login Form with Spring Security>>, with a difference that you need specify the URL for the custom login view.

Reviewer 2: The two Vale reported errors below (preposition and spelling) would be good to address.

.frontend/login-view.ts
[source, typescript]
----
@customElement('login-view')
export class LoginView extends LitElement implements AfterEnterObserver {

  @property({type: Boolean})
  private error = false;

  @property({type: Boolean})
  private open = true;

  // the url to redirect to after a successful login
  private returnUrl = '/';

  private onSuccess: (result:LoginResult) => void;

  constructor(){
    super();
    this.onSuccess = () => {
      Router.go(this.returnUrl);
    };
  }

  render() {
    return html`
      <vaadin-login-overlay
        ?opened="${this.open}" 
        .error=${this.error}
        @login="${this.login}">    
      </vaadin-login-overlay>
    `;
  }

  async login(event: CustomEvent): Promise<LoginResult> {
    this.error = false;
    // use the Vaadin provided login helper method to obtain the LoginResult
    const result = await login(event.detail.username, event.detail.password);
    this.error = result.error;
  
    if (!result.error) {
      this.onSuccess(result);
    }

    return result;
  }

  onAfterEnter(location: RouterLocation) {
    this.returnUrl = location.redirectFrom || this.returnUrl;
  }
}
----

> After the login view is defined, you should also define a route for the login view component, e.g., in the `index.ts` file. Note, the `path` for the login view component should be the same as what you defined in `SecurityConfig` for `http.formLogin().loginPage()`.

Reviewer 2: Also the following Vale complaints could be addressed to comply with the style guidelines.
https://vaadin.com/docs/v14/contributing/documentation-style-guidelines.html#latin-abbreviations-i-e-e-g-etc
There are also a few occurrences of `i.e.` below.

---
title: Adding a Custom Login Form with Spring Security
order: 91
layout: page
---

ifdef::env-github[:outfilesuffix: .asciidoc]

= Adding a Custom Login Form with Spring Security

Instead of using the <<adding-login-form-with-spring-security#, default Spring login form>>, you may want to use your own customized form. This article describes how to do that.

== Dependencies and Server Configuration

The dependencies and server configuration is the same as described in <<adding-login-form-with-spring-security#, Adding a Login Form with Spring Security>>, with a difference that you need to specify the URL of the custom login view.

.SecurityConfig.java

> [source,java]

Reviewer: ```suggestion
[source, java]
```

Author: done

----
@EnableWebSecurity
@Configuration
public class SecurityConfig extends WebSecurityConfigurerAdapter {

  @Override
  protected void configure(HttpSecurity http) throws Exception {
    // Vaadin already handles csrf.

>     http.csrf().disable();

Reviewer 3: Should we do some specific ignore instead of disable, perhaps?

https://github.com/vaadin/flow/issues/9202

Author: done, updated to 
```
// Ignore the login processing url and vaadin endpoint calls
http.csrf().ignoringAntMatchers("/login", "/connect/**");
```

    // specify the URL of the login view, the value of the parameter
    // is the defined route for the login view component.
    http.formLogin().loginPage("login");
  }  
}
----

== Custom login view

The easiest way to make a login view is to use the `<vaadin-login-overlay>` component.
Vaadin provides a `login` helper method for Spring Security based authentication that you can use for the login action.

.frontend/login-view.ts
[source, typescript]
----
@customElement('login-view')
export class LoginView extends LitElement implements AfterEnterObserver {

  @property({type: Boolean})
  private error = false;

  @property({type: Boolean})
  private open = true;

  // the url to redirect to after a successful login
  private returnUrl = '/';

>   private onSuccess: (result:LoginResult) => void;

Reviewer: ```suggestion
  private onSuccess: (result: LoginResult) => void;
```

Author: done

  constructor(){
    super();
    this.onSuccess = () => {
      Router.go(this.returnUrl);
    };
  }

  render() {
    return html`
      <vaadin-login-overlay
        ?opened="${this.open}" 
        .error=${this.error}
        @login="${this.login}">    
      </vaadin-login-overlay>
    `;
  }

  async login(event: CustomEvent): Promise<LoginResult> {
    this.error = false;
    // use the Vaadin provided login helper method to obtain the LoginResult

>     const result = await login(event.detail.username, event.detail.password);

Reviewer 3: Would be nice to have the import line for `login` in this example snippet, so that it is known where it is coming from.

Author: good point, added the import line

    this.error = result.error;
  
    if (!result.error) {
      this.onSuccess(result);
    }

    return result;
  }

  onAfterEnter(location: RouterLocation) {
    this.returnUrl = location.redirectFrom || this.returnUrl;
  }
}
----
After the login view is defined, you should also define a route for the login view component, for example in the `index.ts` file. Note, the `path` for the login view component should be the same as what you defined in `SecurityConfig` for `http.formLogin().loginPage()`.
.frontend/index.ts
[source, typescript]
----
const routes = [
  {
    path: '/login',
    component: 'login-view'
  },
  // more routes
}
----
== Redirect the user to the login view
To protect a view from unauthenticated users, that is, redirect unauthenticated users to the login view, you can use route action.

.frontend/index.ts
[source, typescript]
----
const isUserLoggedIn = function() {
  return localStorage.getItem('loggedIn') === 'true';
}

const routes = [
  ...
  {
    path: '/my-view',
    action: async (_: Router.Context, commands: Router.Commands) => {

>       if (!await isUserLoggedIn()) {

Reviewer: async and await are unnecessary here with the current implementation.

        return commands.redirect('/login');
      }
      return undefined;
    },
    component: 'my-view'
  }
  ...
}
----
You can also add the route action to the parent layout, so that all the child views are protected.

.frontend/index.ts
[source, typescript]
----
const routes = [
  ...
  {

>     path: '/main-layout',

Reviewer 3: The main layout is supposed to apply to the entire application, maybe this `path` should be `'/'` accordingly.

Author: good catch, fixed

    action: async (_: Router.Context, commands: Router.Commands) => {

>       if (!await isUserLoggedIn()) {

Reviewer: async and await are unnecessary here with the current implementation.

Author: good catch, done

        return commands.redirect('/login');
      }
      return undefined;
    },
    component: 'main-layout',
    children: [
      ...
    ]
  }
  ...
}
----
The `isUserLoggedIn()` method in the above code examples uses

> a `loggedIn` boolean stored in the https://vaadin.com/components/browse[localstorage] to check if the user is logged in. The boolean needs to be reset when logging out.

Reviewer 3: The link is misleading, `localStorage` is not a Vaadin component, but a web platform feature. Let’s remove the link, or link MDN instead: https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage

Author: yep, a bad copy paste mistake, fixed.

> a `loggedIn` boolean stored in the https://vaadin.com/components/browse[localstorage] to check if the user is logged in. The boolean needs to be reset when logging out.

Reviewer: Why is there a link to https://vaadin.com/components/browse here? Should it maybe link to MDN docs about localStorage?

Reviewer: Also "localstorage" should be written as "localStorage".

Reviewer: It would be good to include some explanation of why we're using a locally stored boolean for checking logged in status (so that it works offline too) and that this is not a security issue because it is only used to decide which views will be shown from the views that are already available in the client. And include some note that the content included directly in the TS views files (behind authentication) should not be considered secure (unless the whole view is loaded dynamically after authenticating to the server) and all secure data needs to be fetched via endpoints where the server makes sure the user is authenticated.

Or do we have this kind of explanation somewhere else already?

Author: The link was a bad copy past mistake, fixed.
Add a brief explanation why it was used. Note this is for V18, which doesn't have all the offline feature we are developing, so prefer not to expose much to it. For it to work better offline, would also need to use some timestamp, will update this once offline is to be merged to master. 
Regarding security, I think it's explained in this [page](https://vaadin.com/docs/v18/flow/typescript/configuring-security.html).

Reviewer: @haijian-vaadin that page seems to be mainly about endpoint security. It doesn't tell anything about the security of content in the TS view files.

My main concern is that someone sees our example for adding authentication to a view and then has some information in the view (not fetched via endpoint, but hardcoded into the view) that is not supposed to be seen by non-authenticated users. So it should be explained somewhere that this way is not secure. Because the user could just set the loggedIn value in localStorage themselves to see all the views. But it will probably take some time to write a meaningful section about this and could be left outside the scope of this PR imo.

Author: I see, very good point, this could be some best practices article/tutorial with security. Let's do it in another article.

== Logout
There are two ways to implement logout. You could either:

Let Spring Security handle Logging out. How it works is that it handles HTTP GET requests to `/logout` and redirects to `/login?logout` in response.
For that a "Logout" button should be an regular <a> tag, that is,

> `<a href="/logout" router-ignore>Log out</a>`. The downside is that it requires a full page load.

Reviewer: I don't think this works with the localStorage implementation anymore, since this way it would not clear the "loggedIn" status and you would still be able to see the views even after navigating to "/logout". I think this part can probably be just removed. Or if we want to keep this as an option, we would probably need to add some code that would update the "loggedIn" status in localStorage by checking it from the server on page load (reload) **if** the browser is online.

Reviewer: If this is removed, also update the "There are two ways to implement logout. You could either:" above since there would only be one example anymore.

Author: good catch, removed the content.

}
----
The `isUserLoggedIn()` method in the above code examples uses

> a `loggedIn` boolean stored in the https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage[localStorage] to check if the user is logged in. The boolean needs to be reset when logging out. Use localStorage allows navigating to sub views without always having to check authentication from the backend. 

Reviewer: ```suggestion
a `loggedIn` boolean stored in the https://developer.mozilla.org/en-US/docs/Web/API/Window/localStorage[localStorage] to check if the user is logged in. The boolean needs to be reset when logging out. Using localStorage allows navigating to sub views without having to check authentication from the backend on every navigation.
```

Author: done
